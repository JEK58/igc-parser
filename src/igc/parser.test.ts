import IGCParser from './parser';

import fs = require('fs');

describe('IGCParser', () => {
  let parser: IGCParser;

  beforeEach(() => {
    parser = new IGCParser();
  });

  describe('parse()', () => {
    it('parses valid IGC files', () => {
      let filename = `${__dirname}/../../fixtures/2017-07-15-lev/1G_77fv6m71.igc`;
      let content = fs.readFileSync(filename, 'utf8');
      let result = IGCParser.parse(content);

      expect(result.aRecord).toEqual({
        manufacturer: 'LXNAV',
        loggerId: '6M7',
        numFlight: 1,
        additionalData: null,
      });

      expect(result.date).toEqual('2017-07-15');
      expect(result.pilot).toEqual('Florian Graf');
      expect(result.copilot).toBeUndefined();
      expect(result.gliderType).toEqual('ASW 19');
      expect(result.registration).toEqual('D-2019');
      expect(result.callsign).toEqual('1G');
      expect(result.competitionClass).toEqual('Club');

      expect(result.fixes.length).toEqual(4047);
      expect(result.fixes[0]).toMatchSnapshot();
      expect(result.fixes[1234]).toMatchSnapshot();
      expect(result.fixes[2042]).toMatchSnapshot();
      expect(result.fixes[4046]).toMatchSnapshot();
    });

    it('throws if HFDTE is missing', () => {
      let lines = [
        'ALXV6M7FLIGHT:1',
        'HFGIDGLIDERID:D-2019',
      ];

      expect(() => IGCParser.parse(lines.join('\n'))).toThrowErrorMatchingSnapshot();
    });

    it('handles UTC midnight correctly', () => {
      let result = IGCParser.parse([
        'ALXV6M7',
        'HFDTE150717',
        'B1926130000000N00000000EA0000000000',
        'B2300210000000N00000000EA0000000000',
        'B0001550000000N00000000EA0000000000',
        'B2359590000000N00000000EA0000000000',
        'B0242420000000N00000000EA0000000000',
        'B0659000000000N00000000EA0000000000',
        'B2200000000000N00000000EA0000000000',
        'B0100000000000N00000000EA0000000000',
      ].join('\n'));

      let fixTimestamps = result.fixes.map(fix => (new Date(fix.timestamp)).toISOString());
      expect(fixTimestamps).toEqual([
        '2017-07-15T19:26:13.000Z',
        '2017-07-15T23:00:21.000Z',
        '2017-07-16T00:01:55.000Z',
        '2017-07-15T23:59:59.000Z',
        '2017-07-16T02:42:42.000Z',
        '2017-07-16T06:59:00.000Z',
        '2017-07-16T22:00:00.000Z',
        '2017-07-17T01:00:00.000Z',
      ]);
    });
  });

  describe('parseARecord()', () => {
    it('parses valid HFDTE headers', () => {
      expect(parser['parseARecord']('AFLA1BG')).toEqual({
        manufacturer: 'Flarm',
        loggerId: '1BG',
        numFlight: null,
        additionalData: null,
      });

      expect(parser['parseARecord']('AMMMNNN:TEXT STRING')).toEqual({
        manufacturer: 'MMM',
        loggerId: 'NNN',
        numFlight: null,
        additionalData: 'TEXT STRING',
      });

      expect(parser['parseARecord']('ALXV6M7FLIGHT:1')).toEqual({
        manufacturer: 'LXNAV',
        loggerId: '6M7',
        numFlight: 1,
        additionalData: null,
      });

      expect(parser['parseARecord']('ALXNNE2FLIGHT:42')).toEqual({
        manufacturer: 'LX Navigation',
        loggerId: 'NE2',
        numFlight: 42,
        additionalData: null,
      });
    });

    it('throws for invalid records', () => {
      expect(() => parser['parseARecord']('')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseARecord']('HFDTE')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseARecord']('HFDTEXXXXXX')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseARecord']('HFDTE1234?6')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('parseDateHeader()', () => {
    it('parses valid HFDTE headers', () => {
      expect(parser['parseDateHeader']('HFDTE010180')).toEqual({ date: '1980-01-01' });
      expect(parser['parseDateHeader']('HFDTE150717')).toEqual({ date: '2017-07-15' });
      expect(parser['parseDateHeader']('HFDTE311279')).toEqual({ date: '2079-12-31' });
    });

    it('throws for invalid records', () => {
      expect(() => parser['parseDateHeader']('')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseDateHeader']('HFDTE')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseDateHeader']('HFDTEXXXXXX')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseDateHeader']('HFDTE1234?6')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('parsePilot()', () => {
    it('parses valid PLT headers', () => {
      expect(parser['parsePilot']('HFPLTPILOT:')).toEqual('');
      expect(parser['parsePilot']('HFPLTPilotincharge:Mike Havener')).toEqual('Mike Havener');
      expect(parser['parsePilot']('HFPLTPILOT:BEN BRAND')).toEqual('BEN BRAND');
      expect(parser['parsePilot']('HFPLTPILOT:foo:bar')).toEqual('foo:bar');
      expect(parser['parsePilot']('HFPLTPILOTINCHARGE:YUTA_ISHIGURO')).toEqual('YUTA ISHIGURO');
      expect(parser['parsePilot']('HFPLT Pilot             :Gisela Weinreich  ')).toEqual('Gisela Weinreich');
      expect(parser['parsePilot']('HFPLT:John Doe')).toEqual('John Doe');
      expect(parser['parsePilot']('HOPLTJohn Doe')).toEqual('John Doe');
    });

    it('throws for invalid records', () => {
      expect(() => parser['parsePilot']('')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parsePilot']('HXPLT:John Doe')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parsePilot']('HFPLX:John Doe')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('parseCopilot()', () => {
    it('parses valid CM2 headers', () => {
      expect(parser['parseCopilot']('HFCM2CREW2:')).toEqual('');
      expect(parser['parseCopilot']('HFCM2Crew 2:Mike Havener')).toEqual('Mike Havener');
      expect(parser['parseCopilot']('HFCM2CREW2:BEN BRAND')).toEqual('BEN BRAND');
      expect(parser['parseCopilot']('HFCM2CREW2:foo:bar')).toEqual('foo:bar');
      expect(parser['parseCopilot']('HFCM2CREW2:YUTA_ISHIGURO')).toEqual('YUTA ISHIGURO');
      expect(parser['parseCopilot']('HFCM2 Copilot           :Gisela Weinreich  ')).toEqual('Gisela Weinreich');
      expect(parser['parseCopilot']('HFCM2:John Doe')).toEqual('John Doe');
      expect(parser['parseCopilot']('HOCM2John Doe')).toEqual('John Doe');
    });

    it('throws for invalid records', () => {
      expect(() => parser['parseCopilot']('')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseCopilot']('HXCM2:John Doe')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseCopilot']('HFCM1:John Doe')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('parseBRecord()', () => {
    it('parses valid B records', () => {
      parser['dateHeader'] = { date: '2017-02-03' };

      expect(parser['parseBRecord']('B1026555103888N00703115EA0065700751')).toEqual({
        timestamp: 1486117615000,
        time: '10:26:55',
        latitude: 51.0648,
        longitude: 7.051916666666667,
        valid: true,
        pressureAltitude: 657,
        gpsAltitude: 751,
        extensions: {},
        fixAccuracy: null,
        enl: null,
      });

      expect(parser['parseBRecord']('B0000005103888N00703115EV0000000000')).toEqual({
        timestamp: 1486080000000,
        time: '00:00:00',
        latitude: 51.0648,
        longitude: 7.051916666666667,
        valid: false,
        pressureAltitude: null,
        gpsAltitude: null,
        extensions: {},
        fixAccuracy: null,
        enl: null,
      });
    });

    it('throws for missing "dateHeader"', () => {
      expect(() => parser['parseBRecord']('B1026555103888N00703115EA0065700751')).toThrowErrorMatchingSnapshot();
    });

    it('throws for invalid records', () => {
      parser['dateHeader'] = { date: '2017-02-03' };

      expect(() => parser['parseBRecord']('')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseBRecord']('B10a6555103888N00703115EA0065700751')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseBRecord']('B1026555103888X00703115EA0065700751')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseBRecord']('B102655510Z888N00703115EA0065700751')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseBRecord']('B1026555103888N00703115NA0065700751')).toThrowErrorMatchingSnapshot();
    });
  });

  describe('parseIRecord()', () => {
    it('parses valid I records', () => {
      expect(parser['parseIRecord']('I013638FXA')).toEqual([
        { code: 'FXA', start: 36, length: 3 },
      ]);

      expect(parser['parseIRecord']('I043638FXA3940SIU4143ENL4446MOP')).toEqual([
        { code: 'FXA', start: 36, length: 3 },
        { code: 'SIU', start: 39, length: 2 },
        { code: 'ENL', start: 41, length: 3 },
        { code: 'MOP', start: 44, length: 3 },
      ]);

      expect(parser['parseIRecord']('I083638FXA3941ENL4246TAS4751GSP5254TRT5559VAT6063OAT6467ACZ')).toEqual([
        { code: 'FXA', start: 36, length: 3 },
        { code: 'ENL', start: 39, length: 3 },
        { code: 'TAS', start: 42, length: 5 },
        { code: 'GSP', start: 47, length: 5 },
        { code: 'TRT', start: 52, length: 3 },
        { code: 'VAT', start: 55, length: 5 },
        { code: 'OAT', start: 60, length: 4 },
        { code: 'ACZ', start: 64, length: 4 },
      ]);
    });

    it('throws for invalid records', () => {
      expect(() => parser['parseIRecord']('')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseIRecord']('I023638FXA')).toThrowErrorMatchingSnapshot();
      expect(() => parser['parseIRecord']('I0136FXA38')).toThrowErrorMatchingSnapshot();
    });
  });
});
