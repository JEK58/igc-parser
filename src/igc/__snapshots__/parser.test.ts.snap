// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IGCParser parse() parses valid IGC files 1`] = `
Object {
  "enl": 40,
  "extensions": Object {
    "ACZ": "100",
    "ENL": "040",
    "FXA": "060",
    "GSP": "00051",
    "OAT": "2400",
    "TAS": "00000",
    "TRT": "650",
    "VAT": "00010",
  },
  "gpsAltitude": 49,
  "latitude": 51.0107,
  "longitude": 7.010066666666667,
  "pressureAltitude": -42,
  "time": "10:18:26",
  "timestamp": 1500113906000,
  "valid": true,
}
`;

exports[`IGCParser parse() parses valid IGC files 2`] = `
Object {
  "enl": 41,
  "extensions": Object {
    "ACZ": "090",
    "ENL": "041",
    "FXA": "070",
    "GSP": "06662",
    "OAT": "2010",
    "TAS": "24341",
    "TRT": "60-",
    "VAT": "01290",
  },
  "gpsAltitude": 873,
  "latitude": 50.84336666666667,
  "longitude": 6.4120333333333335,
  "pressureAltitude": 773,
  "time": "11:37:45",
  "timestamp": 1500118665000,
  "valid": true,
}
`;

exports[`IGCParser parse() parses valid IGC files 3`] = `
Object {
  "enl": 40,
  "extensions": Object {
    "ACZ": "120",
    "ENL": "040",
    "FXA": "070",
    "GSP": "90033",
    "OAT": "2190",
    "TAS": "89120",
    "TRT": "54-",
    "VAT": "00740",
  },
  "gpsAltitude": 478,
  "latitude": 50.60658333333333,
  "longitude": 6.79495,
  "pressureAltitude": 378,
  "time": "12:28:38",
  "timestamp": 1500121718000,
  "valid": true,
}
`;

exports[`IGCParser parse() parses valid IGC files 4`] = `
Object {
  "enl": 40,
  "extensions": Object {
    "ACZ": "090",
    "ENL": "040",
    "FXA": "060",
    "GSP": "00153",
    "OAT": "2610",
    "TAS": "00000",
    "TRT": "30-",
    "VAT": "00010",
  },
  "gpsAltitude": 50,
  "latitude": 51.0137,
  "longitude": 7.007866666666667,
  "pressureAltitude": -40,
  "time": "14:39:10",
  "timestamp": 1500129550000,
  "valid": true,
}
`;

exports[`IGCParser parse() throws if HFDTE is missing 1`] = `"Missing HFDTE record"`;

exports[`IGCParser parseARecord() throws for invalid records 1`] = `"Invalid A record: "`;

exports[`IGCParser parseARecord() throws for invalid records 2`] = `"Invalid A record: HFDTE"`;

exports[`IGCParser parseARecord() throws for invalid records 3`] = `"Invalid A record: HFDTEXXXXXX"`;

exports[`IGCParser parseARecord() throws for invalid records 4`] = `"Invalid A record: HFDTE1234?6"`;

exports[`IGCParser parseBRecord() throws for invalid records 1`] = `"Invalid B record: "`;

exports[`IGCParser parseBRecord() throws for invalid records 2`] = `"Invalid B record: B10a6555103888N00703115EA0065700751"`;

exports[`IGCParser parseBRecord() throws for invalid records 3`] = `"Invalid B record: B1026555103888X00703115EA0065700751"`;

exports[`IGCParser parseBRecord() throws for invalid records 4`] = `"Invalid B record: B102655510Z888N00703115EA0065700751"`;

exports[`IGCParser parseBRecord() throws for invalid records 5`] = `"Invalid B record: B1026555103888N00703115NA0065700751"`;

exports[`IGCParser parseBRecord() throws for missing "dateHeader" 1`] = `"Missing HFDTE record before first B record"`;

exports[`IGCParser parseDateHeader() throws for invalid records 1`] = `"Invalid HFDTE record: "`;

exports[`IGCParser parseDateHeader() throws for invalid records 2`] = `"Invalid HFDTE record: HFDTE"`;

exports[`IGCParser parseDateHeader() throws for invalid records 3`] = `"Invalid HFDTE record: HFDTEXXXXXX"`;

exports[`IGCParser parseDateHeader() throws for invalid records 4`] = `"Invalid HFDTE record: HFDTE1234?6"`;

exports[`IGCParser parseIRecord() throws for invalid records 1`] = `"Invalid I record: "`;

exports[`IGCParser parseIRecord() throws for invalid records 2`] = `"Invalid I record: I023638FXA"`;

exports[`IGCParser parseIRecord() throws for invalid records 3`] = `"Invalid I record: I0136FXA38"`;
